### PR TITLE
16723 Update NR affiliation/error notifications

### PR DESCRIPTION
**Pull request by chenhongjing**
*Issue #:*
https://github.com/bcgov/entity/issues/16723

*Description of changes:*
- Show green background in table for 4 seconds if NR was successfully affiliated
- Show toast message for 4 seconds if NR was successfully affiliated
- Show error dialog if NR couldn't be added
- Update unit tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-auth license (Apache 2.0).

### CODE REVIEW

**Comment by seeker25** [comment]
I put as do not merge for now, just pushing out a prod release today. Will take that label off after.

**Comment by severinbeauvais** [comment]
Please check with Sarah whether these changes can be deployed to Prod asap or whether they need to wait and go out with the other changes later.

**Comment by pwei1018** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2381-11958gzo.web.app

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (chenhongjing) [inline auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:345]: I think this will fix an issue related to filters. If we added a new row when some column was hidden, the table updated with all the columns shown, which was not consistent with the `Columns to Show` checkbox.

image

File: auth-web/src/components/datatable/BaseVDataTable.vue
Comment (chenhongjing) [inline auth-web/src/components/datatable/BaseVDataTable.vue:65]: Sometimes the component doesn't re-render when `highlightIndex` changes. To force it updates, I make key change at the same time so that the component can be refreshed every time.

I use `highlightIndex+2` to keep non-zero number so that it avoids duplicate key warnings.

Comment (severinbeauvais) [reply auth-web/src/components/datatable/BaseVDataTable.vue:65]: Yes, "key" is how the component knows whether to re-render items.

It sounds like your fix will ensure that all items have unique keys, including the new "first item" (where key=0).

Another thing that might work is to use `:key="Math.random()" so they're always different, but this might re-render the entire table when an item is added/changed (due to reactivity), and ideally you would only add/update the new item.

In short, I think this is OK, but could you add a comment above this to describe what/why you're doing this?

**Comment by pwei1018** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2381-11958gzo.web.app

File: auth-web/src/components/datatable/BaseVDataTable.vue
Comment (chenhongjing) [reply auth-web/src/components/datatable/BaseVDataTable.vue:65]: To follow up my comments above, the thing I did also make vue re-render the entire table because every key will change when `highlightIndex` changes. I've tested `:key="Math.random()"` also works. Do you think I had better use `Math.random()` instead? 👀

Comment (severinbeauvais) [reply auth-web/src/components/datatable/BaseVDataTable.vue:65]: Haha, you exactly understand what's going on -- awesome 👍 

Is there any way to assign a new key to an item that is inserted at the top of the table?

Comment (severinbeauvais) [reply auth-web/src/components/datatable/BaseVDataTable.vue:65]: Ideally we wouldn't re-render the entire table when a new item is added.

Perhaps the keys could be assigned in code not in script?

Comment (chenhongjing) [reply auth-web/src/components/datatable/BaseVDataTable.vue:65]: What does it mean, in code not in script? :frowning:

Comment (severinbeauvais) [reply auth-web/src/components/datatable/BaseVDataTable.vue:65]: In the data array itself, where it can be done once for each item, instead of in script, where it's computed per item every time it re-renders.

Maybe in script is OK, if the list is rendered just once. This will need testing because I don't know off-hand whether the entire list is always re-rendered, or new/changed items are inserted. I'd be OK with you using the random function if you want to try it.

Comment (chenhongjing) [reply auth-web/src/components/datatable/BaseVDataTable.vue:65]: I'll keep my solution if there's no better way, considering that random function makes the key unstable and unpredictable.

Comment (severinbeauvais) [reply auth-web/src/components/datatable/BaseVDataTable.vue:65]: Is there any way to not re-render the entire table?

**Comment by seeker25** [comment]
Previous to merging this table, please test it with orgs with a high number of affiliations. I'll grab you an org id in a sec

**Comment by chenhongjing** [comment]
I've updated the key for the table. It's more straightforward and also works in this scenario. The loading time will be a little longer with a large number of affiliations. I think it's okay. Please help check if it's acceptable to you. 👀

File: auth-web/src/components/datatable/BaseVDataTable.vue
Comment (severinbeauvais) [reply auth-web/src/components/datatable/BaseVDataTable.vue]: Where do we stand with this?

File: auth-web/src/components/auth/manage-business/AddNameRequestForm.vue
Comment (severinbeauvais) [inline auth-web/src/components/auth/manage-business/AddNameRequestForm.vue:178]: @marzmehr FYI ^^

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (severinbeauvais) [inline auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:171]: I think should be `number` (type) not `Number` (class).

File: auth-web/src/components/auth/manage-business/EntityManagement.vue
Comment (severinbeauvais) [inline auth-web/src/components/auth/manage-business/EntityManagement.vue:526]: Theoretically, we will never "not find" the name request, because we will be using a search feature and selecting one of the results. But it's possible that the phone or email don't match.

This is a future problem / different ticket. The change here is OK. No more is neede atm.

File: auth-web/tests/unit/components/EntityManagement.spec.ts
Comment (severinbeauvais) [inline auth-web/tests/unit/components/EntityManagement.spec.ts:219]: This doesn't work, as the setTimeout function is called 5 seconds after this test ends.

One alternative is to call a "done" function in your setTimeout function.

Another way is to "await sleep(5000)" before your expect statements. (I don't know if sleep() or Sleep() is declared in this repo. If not, just add it to utils/.)

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (chenhongjing) [reply auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:171]: Using `number` will cause an error here.
image

**Review by severinbeauvais — COMMENTED** [review]
This is basically acceptable to me but there are few issues left to resolve, including chatting with Sarah Elzinga (or Dave McK) on whether these changes can be merged to main branch (and thus deployed to Test and Prod soon). If not then you should merge these changes to the `feature-business-registry-dashboard-updates` branch (will be merged into main later).

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (severinbeauvais) [reply auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:171]: OK, this is probably due to Composition API syntax. Now I know.

Would it better to default highlightIndex to something instead (maybe even NaN)?

**Comment by pwei1018** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2381-11958gzo.web.app

**Comment by pwei1018** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2381-11958gzo.web.app

**Comment by chenhongjing** [comment]
If there's no more comments, I'll merge into `feature-business-registry-dashboard-updates` branch today. 👀

File: auth-web/src/components/datatable/BaseVDataTable.vue
Comment (severinbeauvais) [reply auth-web/src/components/datatable/BaseVDataTable.vue]: @chenhongjing What's the status of this issue/conversation?

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (severinbeauvais) [reply auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:171]: @chenhongjing Would it make more sense to assign a default number (or NaN)?

File: auth-web/tests/unit/components/EntityManagement.spec.ts
Comment (severinbeauvais) [reply auth-web/tests/unit/components/EntityManagement.spec.ts]: This is fixed. Resolving.

**Comment by severinbeauvais** [comment]
> If there's no more comments, I'll merge into `feature-business-registry-dashboard-updates` branch today. 👀

Don't merge yet: you don't have any approvals, and you have outstanding conversations, and your rebase didn't work correctly (too many files in your diff).

File: auth-web/src/components/datatable/BaseVDataTable.vue
Comment (chenhongjing) [reply auth-web/src/components/datatable/BaseVDataTable.vue]: Resolved!

**Comment by pwei1018** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2381-11958gzo.web.app

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (chenhongjing) [reply auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue]: Sorry, I missed this comment. I set it as -1 by default now.

**Comment by pwei1018** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2381-11958gzo.web.app

**Review by Jxio — APPROVED** [review]
LGTM

**Comment by chenhongjing** [comment]
> > If there's no more comments, I'll merge into `feature-business-registry-dashboard-updates` branch today. 👀
> 
> Don't merge yet: you don't have any approvals, and you have outstanding conversations, and your rebase didn't work correctly (too many files in your diff).

*Update: Rebase issue resolved* 👀

**Review by JazzarKarim — APPROVED** [review]
LGTM

File: auth-web/src/components/auth/manage-business/EntityManagement.vue
Comment (seeker25) [inline auth-web/src/components/auth/manage-business/EntityManagement.vue:485]: type?

Comment (chenhongjing) [reply auth-web/src/components/auth/manage-business/EntityManagement.vue:485]: Resolved in #2385

**Pull request merged**
